File: foundry/py-service/src/widget_svc/api/route.py
"""Reading a request's route template."""

from __future__ import annotations

from starlette.types import Scope

#: The route label used when a request matched no registered route.
#:
#: It exists to bound cardinality. Labelling an unrouted request with its
#: concrete path would let anyone create unbounded series by sending requests to
#: random URLs.
UNMATCHED_ROUTE = "<unmatched>"


def resolve_route(scope: Scope) -> str:
    """Report the route template a request was dispatched to.

    Starlette records the matched route on the scope as it dispatches, so this
    is only meaningful *after* the request has been handled. That is why the
    in-flight gauge is not labelled by route: at the moment a request starts,
    nobody knows which route it belongs to. OpenTelemetry's own convention for
    ``http.server.active_requests`` omits the route for the same reason.

    Re-running the router's matching to learn the route early was tried and
    rejected: FastAPI wraps included routers in a private type whose shape
    changes between releases.
    """
    route = scope.get("route")
    path: str | None = getattr(route, "path", None)
    return path if path else UNMATCHED_ROUTE

File: go-service/internal/transport/http/route.go
package http

import (
	"net/http"
	"strings"
)

// UnmatchedRoute is the route label used when a request matched no registered
// route.
//
// It exists to bound cardinality. Labelling an unrouted request with its
// concrete path would let anyone create unbounded series by sending requests to
// random URLs.
const UnmatchedRoute = "<unmatched>"

// RouteResolver reports the route template a request will be dispatched to,
// before it is dispatched.
//
// Resolving up front — rather than having the router report the template on the
// way back out — is what lets the in-flight gauge be labelled with the route it
// is holding open, and lets the access log and the metrics agree on one value.
type RouteResolver func(*http.Request) string

// MuxResolver resolves routes by asking the mux itself, which is the same
// lookup ServeMux performs when it dispatches.
func MuxResolver(mux *http.ServeMux) RouteResolver {
	return func(r *http.Request) string {
		_, pattern := mux.Handler(r)
		return routeTemplate(pattern)
	}
}

// routeTemplate turns a registered ServeMux pattern into a route label.
//
// Go 1.22 patterns may carry a method ("GET /widgets/{id}") and may end in the
// exact-match marker ("/healthz{$}"); neither belongs in the label.
func routeTemplate(pattern string) string {
	if pattern == "" {
		return UnmatchedRoute
	}

	// Strip a leading method. The space separates it from the path, and a
	// pattern with no method has no space.
	if method, rest, found := strings.Cut(pattern, " "); found {
		_ = method
		pattern = rest
	}
	pattern = strings.TrimSpace(pattern)
	pattern = strings.TrimSuffix(pattern, "{$}")

	// A bare "/" is the catch-all both muxes register to return problem+json
	// for an unrouted request. A genuine root route would be registered as
	// "/{$}", so this is unambiguous.
	if pattern == "/" {
		return UnmatchedRoute
	}
	if pattern == "" {
		return UnmatchedRoute
	}
	return pattern
}
